nuxeo-runtime/nuxeo-runtime-migration/src/main/java/org/nuxeo/runtime/migration/MigrationServiceImpl.java

```diff
@@ -0,0 +1,286 @@
+                    return null;
+                }
+                throw new IllegalArgumentException("Migration: " + id + " already running step: " + currentStep);
+            }
+            MigrationStateDescriptor stateDescr = descr.states.get(state);
+            if (stateDescr == null) {
+                throw new RuntimeException("Invalid current state: " + state + " for migration: " + id);
+            }
+            if (!stepDescr.fromState.equals(state)) {
+                throw new IllegalArgumentException(
+                        "Invalid step: " + step + " for migration: " + id + " in state: " + state);
+            }
+            String time = String.valueOf(System.currentTimeMillis());
+            kv.put(id + STEP, step);
+            kv.put(id + START_TIME, time);
+            kv.put(id + PING_TIME, time);
+            kv.put(id + PROGRESS_MESSAGE, "");
+            kv.put(id + PROGRESS_NUM, "0");
+            kv.put(id + PROGRESS_TOTAL, "-1");
+            kv.put(id, (String) null);
+            return stepDescr.toState;
+        });
+        if (toState == null) {
+            // already running
+            return;
+        }
+
+        KeyValueStore keyValueStore = getKeyValueStore();
+        MigrationContext migrationContext = new MigrationContext() {
+            @Override
+            public void reportProgress(String message, long num, long total) {
+                keyValueStore.put(id + PROGRESS_MESSAGE, message);
+                keyValueStore.put(id + PROGRESS_NUM, String.valueOf(num));
+                keyValueStore.put(id + PROGRESS_TOTAL, String.valueOf(total));
+            }
+        };
+        Runnable runnable = () -> {
+            migrator.run(migrationContext);
+            atomic(id, kv -> {
+                kv.put(id, toState);
+                kv.put(id + STEP, (String) null);
+                kv.put(id + START_TIME, (String) null);
+                kv.put(id + PING_TIME, (String) null);
+                kv.put(id + PROGRESS_MESSAGE, (String) null);
+                kv.put(id + PROGRESS_NUM, (String) null);
+                kv.put(id + PROGRESS_TOTAL, (String) null);
+            });
+        };
+
+        runnable.run(); // TODO XXX in a thread
+    }
+
+    /**
+     * Execute while setting a lock
+     */
+    protected <T> T atomic(String id, Function<KeyValueStore, T> function) {
+        KeyValueStore kv = getKeyValueStore();
+        String lock = String.valueOf(System.currentTimeMillis());
+        if (!kv.compareAndSet(id + LOCK, null, lock, WRITE_LOCK_TTL)) {
+            throw new RuntimeException("Concurrent write to migration: " + id);
+        }
+        try {
+            return function.apply(kv);
+        } finally {
+            kv.put(id + LOCK, (String) null);
+        }
+    }
+
+    protected void atomic(String id, Consumer<KeyValueStore> consumer) {
+        atomic(id, kv -> {
+            consumer.accept(kv);
+            return null;
+        });
+    }
+
```

Comment:
Would it be worth exposing these atomic methods in the `LockHelper` class ?

Reply:
They're small enough that we don't need to factor them for now, but yes eventually something similar will be helpful. Although here there's a convention on related key names which isn't needed in the general case.